src/libstore/derivations.cc

```diff
@@ -154,31 +155,76 @@ StorePath writeDerivation(Store & store,
 }
 
 
+namespace {
+/* This mimics std::istream to some extent. We use this much smaller implementation
+   instead of plain istreams because the sentry object overhead is too high. */
+struct ParseContext {
+    std::string_view str;
+
+    int peek() const {
+        return str.empty() ? EOF : str[0];
+    }
+
+    int get() {
+        if (str.empty()) return EOF;
+        char c = str[0];
+        str.remove_prefix(1);
+        return c;
+    }
+};
+
+constexpr struct Escapes {
+    char map[256];
+    constexpr Escapes() {
+        for (int i = 0; i < 256; i++) map[i] = (char) (unsigned char) i;
+        map[(int) (unsigned char) 'n'] = '\n';
+        map[(int) (unsigned char) 'r'] = '\r';
+        map[(int) (unsigned char) 't'] = '\t';
+    }
+    char operator[](char c) const { return map[(unsigned char) c]; }
+} escapes;
+}
+
+
 /* Read string `s' from stream `str'. */
-static void expect(std::istream & str, std::string_view s)
+static void expect(ParseContext & str, std::string_view s)
 {
-    for (auto & c : s) {
-        if (str.get() != c)
-            throw FormatError("expected string '%1%'", s);
-    }
+    if (!str.str.starts_with(s))
+        throw FormatError("expected string '%1%'", s);
+    str.str.remove_prefix(s.size());
 }
 
 
 /* Read a C-style string from stream `str'. */
-static std::string parseString(std::istream & str)
+static BackedStringView parseString(ParseContext & str)
 {
-    std::string res;
     expect(str, "\"");
-    int c;
-    while ((c = str.get()) != '"')
-        if (c == '\\') {
-            c = str.get();
-            if (c == 'n') res += '\n';
-            else if (c == 'r') res += '\r';
-            else if (c == 't') res += '\t';
-            else res += c;
+    auto c = str.str.begin(), end = str.str.end();
+    bool escaped = false;
+    for (; c != end && *c != '"'; c++) {
+        if (*c == '\\') {
+            c++;
+            if (c == end)
+                throw FormatError("unterminated string in derivation");
+            escaped = true;
+        }
+    }
+
+    const auto contentLen = c - str.str.begin();
+    const auto content = str.str.substr(0, contentLen);
+    str.str.remove_prefix(contentLen + 1);
+
+    if (!escaped)
+        return content;
+
+    std::string res;
+    res.reserve(content.size());
+    for (auto c = content.begin(), end = content.end(); c != end; c++)
```

Comment:
This `c` shadows the other `c`. Either put the other `c` in a block, or reuse it (ie no `auto` declaration, but an assignment).
Same for `end`.